### DISCUSSION TITLE
Generate migration code from entities

### DISCUSSION THREAD

**Discussion by JUNGYU-RIM1** [SeaQL/sea-orm#645] [Ideas]
## Summary
**sea-orm-cli migrate init** should generate migration files from entities

In this example [setting-up-migration](https://www.sea-ql.org/SeaORM/docs/migration/setting-up-migration), sea-orm only generate **m20220101_000001_create_table.rs** which is a sample migration file. If there are pre-defined entities, it should generate those migration files too(in this case **m20220120_000001_create_post_table.rs**).

In the case of the maintenance phase, generating an alter_table file is also needed. If a certain entity is changed, it should find the create_table file of that entity, and based on that file, the alter_table file should be generated.

## Motivation
Currently, if I use the code-first approach in sea-orm, I should write code to generate tables that fit into my entities. There are several problems with that
- There is the possibility that I mismatch entities and code.
- If I change the entity, I should change the migration code too

So I think that adding this feature will help both error-prune and develop performance

## Additional Information
- [EF in c# support this feature ](https://docs.microsoft.com/en-us/ef/ef6/modeling/code-first/workflows/new-database)
-  [Prisma in Node.js also support similar feature](https://www.prisma.io/docs/concepts/components/prisma-migrate)

**Comment by mojerro** [comment]
Hope has this feature too, just like #332 , I really hope to have a function like `auto_migrate()` which is very convenient when developing. 

Now, my project has 16 tables which are really hard to write migration files by hand, and I need to spend lots of time checking and ensuring the code runs correctly.

**Comment by tyt2y3** [comment]
I think SeaORM prefers a schema-first approach (that's just an opinion).
It's a complex problem to tackle (with ambiguities) and I don't think we will be attempting to generate migration scripts soon.

**Reply by JUNGYU-RIM1** [reply]
@tyt2y3 

First of all, thank you for the great library. I think it is the only library that can be called an ORM in the rust ecosystem.

I thought that seaORM prefers a code-first approach because, in the documentation, the explanation of the documentation seems like seaORM prefers to write the entity first then write generation code. (because it explains first how to write generation code if you have pre-defined entities, then explain the "alternative way" if you don't have pre-defined entities ) I write this proposal because in this case, I have to maintain both entities and generation codes and It is inconvenient compared to other ORM(EF or Prisma).

But If I take the schema-first approach, and write migration code first then let seaORM generate entities based on DB schema, it sounds quite reasonable to me because I just maintain migration code.

If you don't mind, may I ask why the documentation explains creating entities first, then explains writing migration code?  I think if seaORM prefers a schema-first approach, the order of explanation should be reversed (Sorry if I misunderstand something)

**Reply by satyamjay-iitd** [reply]
I had the same doubt, commenting for future reference. Coming from sqlalchemy(Python) my thought process was to:
1) **write** the Model(entities)  and
2) **generate** migration script.
3) run migration

Whereas with seaorm you are supposed to
1) **write** migration script first
2) run migration
3) **generate** entities

**Comment by Huliiiiii** [comment]
Closed in favor of #2889
